exla/lib/exla/backend.ex

```diff
@@ -0,0 +1,192 @@
+    from_binary(out, data)
+  end
+
+  defp number_to_binary(number, type),
+    do: match_types([type], do: <<write!(number, 0)>>)
+
+  @impl true
+  def from_binary(t, binary, _backend_options), do: from_binary(t, binary)
+
+  defp from_binary(t, binary) when is_binary(binary), do: %{t | data: %B{state: binary}}
+  defp from_binary(t, other), do: %{t | data: %B{state: IO.iodata_to_binary(other)}}
+
+  @impl true
+  defdelegate backend_copy(tensor, backend, opts), to: Nx.BinaryBackend
+
+  @impl true
+  defdelegate backend_transfer(tensor, backend, opts), to: Nx.BinaryBackend
+
+  @impl true
+  defdelegate backend_deallocate(tensor), to: Nx.BinaryBackend
+
+  @impl true
+  defdelegate to_binary(t, limit), to: Nx.BinaryBackend
+
+  @impl true
+  defdelegate to_batched_list(out, t, x), to: Nx.BinaryBackend
+
+  @impl true
+  def eye(%{shape: shape, type: type, names: names}, _) do
+    fun = fn ->
+      Nx.eye(shape, type: type, names: names, backend: Nx.Defn.Expr)
+    end
+
+    EXLA.jit(fun, [])
+  end
+
+  @impl true
+  def iota(%{shape: shape, type: type, names: names}, axis, _) do
+    fun = fn ->
+      Nx.iota(shape, type: type, names: names, axis: axis, backend: Nx.Defn.Expr)
+    end
+
+    EXLA.jit(fun, [])
+  end
+
+  @impl true
+  def random_uniform(%{shape: shape, type: type, names: names}, min, max, _) do
+    fun = fn ->
+      Nx.random_uniform(shape, min, max, type: type, names: names, backend: Nx.Defn.Expr)
+    end
+
+    EXLA.jit(fun, [])
+  end
+
+  @impl true
+  def random_normal(%{shape: shape, type: type, names: names}, mu, sigma, _) do
+    fun = fn ->
+      Nx.random_normal(shape, mu, sigma, type: type, names: names, backend: Nx.Defn.Expr)
+    end
+
+    EXLA.jit(fun, [])
+  end
+
+  @impl true
+  def inspect(tensor, inspect_opts) do
+    limit = inspect_opts.limit
+    binary = Nx.to_binary(tensor, if(limit == :infinity, do: [], else: [limit: limit + 1]))
+    Nx.Backend.inspect(tensor, binary, inspect_opts)
+  end
+
+  @impl true
+  def concatenate(out, tensors, axis) do
+    expr_fn = fn tensors ->
+      apply(Nx.Defn.Expr, :concatenate, [out, Tuple.to_list(tensors), axis])
+    end
+
+    EXLA.jit(expr_fn, [List.to_tuple(tensors)])
+  end
+
+  binary_ops =
+    [:add, :subtract, :multiply, :power, :remainder, :divide, :atan2, :min, :max, :quotient] ++
+      [:bitwise_and, :bitwise_or, :bitwise_xor, :left_shift, :right_shift] ++
+      [:equal, :not_equal, :greater, :less, :greater_equal, :less_equal] ++
+      [:logical_and, :logical_or, :logical_xor]
+
+  unary_ops =
+    [:exp, :expm1, :log, :log1p, :logistic, :cos, :sin, :tan] ++
+      [:cosh, :sinh, :tanh, :acos, :asin, :atan, :acosh, :asinh, :atanh] ++
+      [:sqrt, :rsqrt, :cbrt, :erf, :erfc, :erf_inv] ++
+      [:abs, :bitwise_not, :ceil, :conjugate, :floor, :negate, :round, :sign] ++
+      [:count_leading_zeros, :population_count, :real, :imag]
+
+  callbacks =
+    [
+      {:as_type, [:out, :tensor], [:tensor]},
```

Comment:
We could reduce this to `{:as_type, [1]}}`, however I think that would be much more error prone.